laurent22/joplin
# Desktop: Resolves #10549: Add exact search for title and body of notes

Resolves https://github.com/laurent22/joplin/issues/10549

# Summary

On this PR I'm adding the possibility of using exact search for the title and body of notes using double quotes:

`title: "2020-03-06"`
`title: "This is a note title with space" body: "Some other term to search"`
etc

Changes:
1. How the `getTerm` function works to allow search input after double quotes (explained better bellow)
2. Created a function that splits the search input for `title:` and `body:` into phrases and words, escaping the necessary double quotes


## Changes to `getTerm`

`getTerm` is a function responsible for taking a query like `title: something body: another thing` into 
```
[
{ name: 'title', value: 'something' }, 
{ name: 'body', value: 'another thing' }
]
```

The problem was that an input like `title: "something" other` would be parsed as
```
[
{ name: 'title', value: 'something' }, 
{ name: 'text', value: 'other' }
]
```

I changed to associate the value after double quotes with the previous value until it finds another "term" (be it title, body, tag, etc):

```
// behaviour added on this PR
[
{ name: 'title', value: 'something' }, 
{ name: 'title', value: 'other' }
]
```

## Changes in the current behaviour 

As I described above, this PR changes some of the search behaviour. If, for example, the user might make a complex search like so:
`title: "something" -other`

it will be required to change the order of the elements

`-other title: "something"`

to have similar output.


# Testing

I added some automated tests to the new functions, but the feature can also be tested manually:

1. Create a note with "2024-05-06" in the title.
2. Create a note with "2024-06-05" in the title.
3. Search for: title: "2024-06-05"
4. Should show only the note with the exact title
